### PR TITLE
Improve stack usage of all the Handle functions

### PR DESCRIPTION
**Pull request by cdvr1993**
- [X] Description and context for reviewers: one partner, one stranger
- [X] Docs (package doc)
- [X] Entry in CHANGELOG.md

TLDR; Reduce stack usage from the rpc handler function from 2520 bytes to 608 bytes.

**UPDATE:**
We decided to keep the methods with value receivers which increased stack usage a bit (~200 bytes), but still gave us ~1.9KB savings. The only change done was to remove the [10]zap.Field from the call struct and allocate that directly in the method that was using it.

**Original description**
Currently one of our services has the following stack trace:
```
10920
200 google.golang.org/grpc.(*Server).serveStreams.func1.2
888 google.golang.org/grpc.(*Server).handleStream
1000 google.golang.org/grpc.(*Server).processStreamingRPC
728 go.uber.org/yarpc/transport/grpc.(*handler).handle
496 go.uber.org/yarpc/transport/grpc.(*handler).handleUnary
544 go.uber.org/yarpc/transport/grpc.(*handler).handleUnaryBeforeErrorConversion
256 go.uber.org/yarpc/transport/grpc.(*handler).callUnary
544 go.uber.org/yarpc/api/transport.InvokeUnaryHandler
104 go.uber.org/yarpc/api/middleware.unaryHandlerWithMiddleware.Handle
160 go.uber.org/yarpc/internal/inboundmiddleware.unaryChain.Handle
136 go.uber.org/yarpc/internal/inboundmiddleware.unaryChainExec.Handle
2520 go.uber.org/yarpc/internal/observability.(*Middleware).Handle                  # We're fixing this one
136 go.uber.org/yarpc/internal/inboundmiddleware.unaryChainExec.Handle
... (omitting for privacy)
```

We noticed that the Handle function is consuming 2.5KB of stack usage. Although, it is preferred to use stack vs heap, >2KB seems like a problem. This could negatively impact an application if it frequently causes stack expansion. Additionally, it requires more memory to hold it.

The problem was with the *call* struct:

```
type call struct {
	edge    *edge
	extract ContextExtractor
        fields  [10]zapcore.Field

	started   time.Time
	ctx       context.Context
	req       *transport.Request
	rpcType   transport.Type
	direction directionName

	levels *levels
}
```

After calling unsafe.Sizeof(call{}), it shows that it consumes 736 bytes. In this case the compiler decided to give stack space for 3 copies of call:

1. The local variable of Handle().
2. The return value from call{}.
3. One for passing the call to the defer or calling any method on call.

This looks somewhat inefficient so we decided to fix it, to have only one copy at the Handle() level. The required changes were:

1. All methods for "call" must use pointer receiver to avoid allocating stack space.
2. begin() function can receive a pointer to the stack allocated call{} coming from Handle() (in this way we eliminate the copy for the return value).
3. Remove the [10]zap.Field from call{}. These fields are then stored in a buffer which cause the stack allocated at Handle() to be escaped to the heap.
4. Fix all Handle() functions (stream, unary, etc) to follow the same process.

After these changes we can quickly see the differences from the assembly:

From
```
TEXT go.uber.org/yarpc/internal/observability.(*Middleware).Handle(SB) gofile../home/user/go/go-no-monorepo/src/go.uber.org/yarpc/internal/observability/middleware.go
  middleware.go:183     0x62dd4                 4881ec88090000                  SUBQ $0x988, SP
--
TEXT go.uber.org/yarpc/internal/observability.(*Middleware).HandleOneway(SB) gofile../home/user/go/go-no-monorepo/src/go.uber.org/yarpc/internal/observability/middleware.go
  middleware.go:235     0x6374a                 4881ec680c0000                  SUBQ $0xc68, SP
--
TEXT go.uber.org/yarpc/internal/observability.(*Middleware).HandleStream(SB) gofile../home/user/go/go-no-monorepo/src/go.uber.org/yarpc/internal/observability/middleware.go
  middleware.go:255     0x63f15                 4881ec38090000                  SUBQ $0x938, SP
```

To
```
TEXT go.uber.org/yarpc/internal/observability.(*Middleware).Handle(SB) gofile../home/user/go/go-no-monorepo/src/go.uber.org/yarpc/internal/observability/middleware.go
  middleware.go:183     0x5c728                 4881ec50010000                  SUBQ $0x150, SP
--
TEXT go.uber.org/yarpc/internal/observability.(*Middleware).HandleOneway(SB) gofile../home/user/go/go-no-monorepo/src/go.uber.org/yarpc/internal/observability/middleware.go
  middleware.go:241     0x5cf29                 4881ec50010000                  SUBQ $0x150, SP
--
TEXT go.uber.org/yarpc/internal/observability.(*Middleware).HandleStream(SB) gofile../home/user/go/go-no-monorepo/src/go.uber.org/yarpc/internal/observability/middleware.go
  middleware.go:263     0x5d553                 4881ec28010000                  SUBQ $0x128, SP
```

Finally our application reports:

```
8824
200 google.golang.org/grpc.(*Server).serveStreams.func1.2
888 google.golang.org/grpc.(*Server).handleStream
1000 google.golang.org/grpc.(*Server).processStreamingRPC
728 go.uber.org/yarpc/transport/grpc.(*handler).handle
496 go.uber.org/yarpc/transport/grpc.(*handler).handleUnary
544 go.uber.org/yarpc/transport/grpc.(*handler).handleUnaryBeforeErrorConversion
256 go.uber.org/yarpc/transport/grpc.(*handler).callUnary
544 go.uber.org/yarpc/api/transport.InvokeUnaryHandler
104 go.uber.org/yarpc/api/middleware.unaryHandlerWithMiddleware.Handle
160 go.uber.org/yarpc/internal/inboundmiddleware.unaryChain.Handle
136 go.uber.org/yarpc/internal/inboundmiddleware.unaryChainExec.Handle
424 go.uber.org/yarpc/internal/observability.(*Middleware).Handle
136 go.uber.org/yarpc/internal/inboundmiddleware.unaryChainExec.Handle
...
```

424 vs the original 2520.

### CODE REVIEW

**Comment by rabbbit** [comment]
Are there any benchmarks you could run?

File: internal/observability/graph.go [hunk 1]
```diff
@@ -93,9 +93,9 @@ func newGraph(meter *metrics.Scope, logger *zap.Logger, extract ContextExtractor
 	}
 }
 
-// begin starts a call along an edge.
-func (g *graph) begin(ctx context.Context, rpcType transport.Type, direction directionName, req *transport.Request) call {
-	now := _timeNow()
+func (g *graph) begin(
+	ctx context.Context, rpcType transport.Type, direction directionName, req *transport.Request, c *call) {
```
Comment (rabbbit) [inline internal/observability/graph.go:97]: would returning a `*call` instead of `call` not be enough?

File: internal/observability/middleware.go [hunk 2]
```diff
@@ -181,8 +181,12 @@ func applyLogLevelsConfig(dst *levels, src *DirectionalLevelsConfig) {
 
 // Handle implements middleware.UnaryInbound.
 func (m *Middleware) Handle(ctx context.Context, req *transport.Request, w transport.ResponseWriter, h transport.UnaryHandler) error {
-	call := m.graph.begin(ctx, transport.Unary, _directionInbound, req)
-	defer m.handlePanicForCall(call, transport.Unary)
+
+	var call call
+	m.graph.begin(ctx, transport.Unary, _directionInbound, req, &call)
+	// It is safe to pass a pointer to defer because the stack memory doesn't get released so there is no escape to the
+	// heap
```
Comment (rabbbit) [inline internal/observability/middleware.go:188]: I don't understand - is there a guarantee that the call won't get heap allocated anyway?

Separately, what is the concern? That this memory is on a heap, so you're passing a pointer to memory that might get released (?)

File: internal/observability/graph.go [hunk 3]
```diff
@@ -93,9 +93,9 @@ func newGraph(meter *metrics.Scope, logger *zap.Logger, extract ContextExtractor
 	}
 }
 
-// begin starts a call along an edge.
-func (g *graph) begin(ctx context.Context, rpcType transport.Type, direction directionName, req *transport.Request) call {
```
Comment (DheerendraRathor) [inline internal/observability/graph.go:97]: Does this need to be changed? 
Were you able to measure impact after removing zap fields from call and making call method pointers?

File: internal/observability/middleware.go [hunk 4]
```diff
@@ -181,8 +181,12 @@ func applyLogLevelsConfig(dst *levels, src *DirectionalLevelsConfig) {
...
+	defer m.handlePanicForCall(&call, transport.Unary)
```
Comment (rabbbit) [inline internal/observability/middleware.go:189]: is it not easier to change `m.handlePanicForCall` to `c.handlePanicForCall` ?

it doesn't seem to depend on anything in `m`

File: internal/observability/call.go [hunk 5]
```diff
@@ -222,7 +221,8 @@ func (c call) endLogs(
 		return
 	}
 
-	fields := c.fields[:0]
+	backendFields := [10]zapcore.Field{}
+	fields := backendFields[:0]
```
Comment (DheerendraRathor) [inline internal/observability/call.go:225]: nit: group into var
```
var (
    tmp [10]zapcore.Field

    fields = tmp[:0]
)
```

File: internal/observability/graph.go (hunk 3 above)
Comment (rabbbit) [reply internal/observability/graph.go:97]: +1.

File: internal/observability/middleware.go (hunk 4 above)
Comment (cdvr1993) [reply internal/observability/middleware.go:189]: Not sure, I wanted to touch as little as possible the original code.

File: internal/observability/middleware.go (hunk 2 above)
Comment (cdvr1993) [reply internal/observability/middleware.go:188]: is there a guarantee that the call won't get heap allocated anyway?
- No, there is no guarantee if someone changes the code, but as of now it would always be on the stack because its liveness can be determined at compile time.

Separately, what is the concern? That this memory is on a heap, so you're passing a pointer to memory that might get released (?)
- Concern? Performance impact. Heap allocation is more expensive than stack, but other than that there is no bug/issue if it is either on the stack or the heap.

File: internal/observability/graph.go (hunk 1 above)
Comment (cdvr1993) [reply internal/observability/graph.go:97]: It seems it is not the same. What you suggests escapes to the heap:

```
graph.go:113          0x59e3e                 e800000000              CALL 0x59e43                                                            [1:5]R_CALL:runtime.newobject
```

File: internal/observability/graph.go (hunk 3 above)
Comment (cdvr1993) [reply internal/observability/graph.go:97]: if I do zap change in another diff my stack allocated object becomes a heap allocated object. [10]zap.Fields must not be attached to call to avoid that.

File: internal/observability/call.go (hunk 5 above)
Comment (cdvr1993) [reply internal/observability/call.go]: It seems this is not needed. Doing `make([]zap.Field, 0, 10)` still does stack allocation.

**Comment by cdvr1993** [comment]
> Are there any benchmarks you could run?

Added the benchmark code. The result is here:

### No assert
#### Previous
Screenshot 2023-07-21 at 3 26 17 PM

#### Now
Screenshot 2023-07-21 at 3 27 16 PM


### With assert
#### Previous
Screenshot 2023-07-21 at 3 29 59 PM

#### Now
Screenshot 2023-07-21 at 3 29 49 PM

File: internal/observability/graph.go (hunk 3 above)
Comment (DheerendraRathor) [reply internal/observability/graph.go:97]: If you leave graph.begin changes untouched, what is the performance impact? 

 > [10]zap.Fields must not be attached to call to avoid that.
This change is okay and looks good.

File: internal/observability/middleware.go (hunk 2 above)
Comment (DheerendraRathor) [reply internal/observability/middleware.go:188]: So let's keep it on stack then and return call from graph like it was before ? 
Since zap fields are removed, copy cost is going to be low now and it's going to happen only once per call, so it should be okay. 

Can you separately benchmark this after making call & zap changes and if graph changes actually improve performance at all ?

**Comment by DheerendraRathor** [comment]
@cdvr1993 I ran benchmarks with and without graph changes (keeping just call & zap changes) on my laptop. 

Without graph changes:
```
$ go test -bench=BenchmarkMiddlewareHandle -benchmem
goos: darwin
goarch: amd64
pkg: go.uber.org/yarpc/internal/observability
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkMiddlewareHandle-8      2292955               517.4 ns/op           192 B/op          2 allocs/op
PASS
ok      go.uber.org/yarpc/internal/observability        2.240s
```

With graph changes (this PR as-is)
```
$ go test -bench=BenchmarkMiddlewareHandle -benchmem
goos: darwin
goarch: amd64
pkg: go.uber.org/yarpc/internal/observability
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkMiddlewareHandle-8      2327730               515.1 ns/op           192 B/op          2 allocs/op
PASS
ok      go.uber.org/yarpc/internal/observability        2.289s

```

There isn't a significant difference in both.

File: internal/observability/graph.go (hunk 1 above)
Comment (rabbbit) [reply internal/observability/graph.go:97]: I think one *call object escaping to heap is unavoidable - we might be able to over-fit this into our benchmark and temporarily avoid it, but it seems just like it's hard to avoid.

And if heap-ing means a significant code simplification, I'd choose that.

**Comment by cdvr1993** [comment]
@DheerendraRathor Ok, I updated it. It seems the benchmark is faster as you said. Although stack usage is greater:

```
middleware.go:183     0x62d6f                 4881ec60020000          SUBQ $0x260, SP
```

vs

```
  middleware.go:183     0x5c728                 4881ec50010000                  SUBQ $0x150, SP
```

If we want to have more common and readable code removing only fields from call is the way to do it.

If we want to reduce as much as possible stack usage in the middleware function, the original PR is the best. I'll let you to decide.

**Comment by rabbbit** [comment]
@cdvr1993 could you post the final results?

**Comment by DheerendraRathor** [comment]
@cdvr1993 I think there is some miscommunication here between comments. 
I've no objection in converting all call methods to pointer receiver unless there is huge gain in performance with value receivers which I don't think is the case. This will also help you in stack use reduction. 

I had objection in changes to `graph.begin` API which IMO reduced the code readability and maintainability.

**Comment by jronak** [comment]
I can't seem to look at the history of the PR, but what is the difference in the benchmarking numbers between the current changes vs with pointer receivers? 

Thanks for this @cdvr1993!

**Comment by cdvr1993** [comment]
> I can't seem to look at the history of the PR, but what is the difference in the benchmarking numbers between the current changes vs with pointer receivers?
> 
> Thanks for this @cdvr1993!

Negligible. In terms of performance they are pretty much the same. The only advantage would be to reduce stack usage, but most of the usage is gone by removed the zap fields.

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,7 +6,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 =======
 ## [Unreleased]
-- No changes yet.
+- middleware stack usage: remove ~2KB of stack usage from the rpc handler function,
```
Comment (jronak) [inline CHANGELOG.md:9]: nit: is it still 2kb of stack usage reduction?

**Review by jronak — APPROVED** [review]
This looks good. We can go back to pointer receiver methods if needed. Thanks!

**Comment by cdvr1993** [comment]
@jronak 0x988 vs 0x260, so 1.8KB

**Comment by cdvr1993** [comment]
@rabbbit last benchmark results
```
go test ./internal/observability/... -bench . -count=10
goos: linux
goarch: amd64
pkg: go.uber.org/yarpc/internal/observability
cpu: AMD EPYC 7B13
BenchmarkMiddlewareHandle-96             1429688               844.9 ns/op
BenchmarkMiddlewareHandle-96             1443886               828.4 ns/op
BenchmarkMiddlewareHandle-96             1442847               828.4 ns/op
BenchmarkMiddlewareHandle-96             1451000               821.1 ns/op
BenchmarkMiddlewareHandle-96             1442511               833.8 ns/op
BenchmarkMiddlewareHandle-96             1436607               834.3 ns/op
BenchmarkMiddlewareHandle-96             1462096               815.1 ns/op
BenchmarkMiddlewareHandle-96             1425666               816.2 ns/op
BenchmarkMiddlewareHandle-96             1450243               824.1 ns/op
BenchmarkMiddlewareHandle-96             1458702               815.5 ns/op
PASS
ok      go.uber.org/yarpc/internal/observability        20.577s
```

**Review by rabbbit — CHANGES_REQUESTED** [review]
We spoke about this on slack. 

@cdvr1993 could you please:
- prepare a final commit message
- post benchmarks in before/after format
- post `-benchmem` in the benchmarks?

I realize this is likely not going to change the final outcome of this PR, but let's be diligent.

**Comment by cdvr1993** [comment]
@rabbbit done.

**Comment by rabbbit** [comment]
Do you know why the allocations were reduced too?

**Review by rabbbit — APPROVED** [review]

File: internal/observability/call.go [hunk 7]
```diff
@@ -222,7 +221,7 @@ func (c call) endLogs(
 		return
 	}
 
-	fields := c.fields[:0]
+	fields := make([]zapcore.Field, 0, 10)
```
Comment (biosvs) [inline internal/observability/call.go:224]: This is variadic function that accept arbitrary number of filed in `extraLogFields` argument (it works now somehow, because there are no calls with more than one field in extraLogFields, I guess).

So I would say it's better to use smth like:
`fields := make([]zapcore.Field, 0, 9+len(extraLogFields))`

Otherwise one day someone will add another field, and we'll have performance penalty here for expending slice (new allocation + copy).

**Comment by cdvr1993** [comment]
@biosvs updated it. Thanks.

**Review by DheerendraRathor — APPROVED** [review]

**Review by AllenLuUber — APPROVED** [review]
